src/main/php/org/bovigo/vfs/vfsStreamFile.php

```diff
@@ -21,19 +21,19 @@ class vfsStreamFile extends vfsStreamAbstractContent
     /**
      * content of the file
      *
-     * @type  FileContent
+     * @var  FileContent
      */
     private $content;
     /**
      * Resource id which exclusively locked this file
      *
-     * @type  string
+     * @var  string|null
      */
     protected $exclusiveLock;
     /**
      * Resources ids which currently holds shared lock to this file
      *
-     * @type  bool[string]
+     * @var  array<string, bool>
```

Comment:
Is this valid PHPDoc? I have never seen this style used before with any PHP documentation parser/generator.

I would think `bool[]|string[]`. If you wanted more accuracy, `(bool|string)[]`. This form is from the PSR-5 draft and isn't widely supported yet though.

Reply:
It's actually a `bool[]` indexed by `string`. The doc format is [PHP Generics](https://wiki.php.net/rfc/generics), which is still just an RFC but is what phpstan uses to add more detail to known data structures. It's essentially a `<key, value>` comment, in this case. Generics can be used for a lot more than that. There's big projects that have *some* usage of it, but not a lot.
- [phpunit](https://github.com/sebastianbergmann/phpunit/blob/master/src/Util/PHP/AbstractPhpProcess.php#L124)
- [Doctrine](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php#L177)
- phpstan's code is loaded with usage of it

Documenting it using the generics syntax will help static analysis tools ensure data quality, as it should trigger an error if someone ever tries to use a non-string as an index in that array.

Alternatively, it could be `@var bool[] Indexed by string` but it'll be a little less helpful for static analysis.

Reply:
Ah I see! I knew about the Generics syntax, but didn't realize you could apply it to `array`.

Reply:
And yes this is a valid phpdoc, we introduced that I think about 4 years ago ;-)